pkg/ip/ip.go

```diff
@@ -918,3 +919,44 @@ func GetIPFromListByFamily(ipList []net.IP, v4Family bool) net.IP {
 
 	return nil
 }
+
+// AddrFromIP converts a net.IP to netip.Addr using netip.AddrFromSlice, but preserves
+// the original address family. It assumes given net.IP is not an IPv4 mapped IPv6
+// address.
+//
+// The problem behind this is that when we convert the IPv4 net.IP address with
+// netip.AddrFromSlice, the address is interpreted as an IPv4 mapped IPv6 address in some
+// cases.
+//
+// For example, when we do netip.AddrFromSlice(net.ParseIP("1.1.1.1")), it is interpreted
+// as an IPv6 address "::ffff:1.1.1.1". This is because 1) net.IP created with
+// net.ParseIP(IPv4 string) holds IPv4 address as an IPv4 mapped IPv6 address internally
+// and 2) netip.AddrFromSlice recognizes address family with length of the slice (4-byte =
+// IPv4 and 16-byte = IPv6).
+//
+// By using AddrFromIP, we can preserve the address family, but since we cannot distinguish
+// IPv4 and IPv4 mapped IPv6 address only from net.IP value (see #37921 on golang/go) we
+// need an assumption that given net.IP is not an IPv4 mapped IPv6 address.
+func AddrFromIP(ip net.IP) (netip.Addr, bool) {
```

Comment:
Is this necessary? 

If the user is supplying and IPv4 address to `netip.AddrFromSlice` its assumed the user understands that it will want a ipv4 address. The expectation is the user then calls 

```
func (ip Addr) Unmap()
``` 

To unmap the IPv4 address and obtain the IPv4 Addr. 

There is also `Is4In6` for the user to determine if the unmap call is necessary.

Reply:
Hmm. I think this is a temporary gotcha while handling the `net.IP` -> `netip.Addr` transition, and most callers don't know if they're supplying an IPv4 address or IPv6 address. This is partly by design, to avoid all the v4 vs v6 checks scattered throughout the codebase. At a glance it looks to me like you're right that this function could be simplified to something like the following (untested):

```
if addr, ok := netip.AddrFromSlice(ip); !ok {
    return addr, ok
}
return addr.Unmap(), true
```

So then the question is, do we assume that all callers are going to do this exactly like this and get it right (ie don't forget `Unwrap()`, or do we want a one-liner convenience function which does this for us? There's certainly a convenience about making a rule that if you translate between `net` types and `netip` types, we put the conversion functions in `pkg/ip` and you just have to use these converters. That way it becomes easier and easier over time to locate all of the conversions and eventually get rid of them all, while in the mean time having well-tested canonical ways to convert between the types.

Reply:
I also think many callers don't know the address or family of the `net.IP`. That's why we do `ip.To4() != nil` in many places (e.g. in service-related code), and it reduces a few lines of code everyone needs to write and avoid common mistakes (at this point, I and Chris ↓ already made a mistake at least a time). In that sense, it is necessary for my opinion.

BTW, I fixed the code to use `Unmap()`. Thanks for your suggestion!

Reply:
Cool, LGTM.